### PR TITLE
internal alert when sensors client or location is performing outside of normal range (CU-2chw8y0)

### PR DESCRIPTION
**Pull request by susanchung-brave**
so that we can detect and correct issues in the system before it becomes a problem for clients!

test plan:
- [x] physical sensor:
- [x] set interval to be 10 minutes + make stillness alert threshold shorter
    - [x] trigger a bunch of stillness alerts (way past maxStillnessAlerts)
        - [x] confirm that it logs to Sentry
    - [x] trigger maxStillnessAlerts + 1 stillness alerts
        - [x] confirm that it logs to Sentry
    - [x] trigger exactly maxStillnessAlerts stillness alerts
        - [x] confirm that it doesn't log to Sentry
    - [x] trigger like 1 stillness alert
        - [x] confirm that it doesn't log to Sentry
    - [x] don't trigger a stillness alert
        - [x] confirm it doesn't log to Sentry
- [x] do same series of tests as above but for a different interval (like 5 minutes to make my life easier)
- [x] POST requests:
- [x] set interval to be 10 minutes
    - [x] trigger a bunch of stillness alerts (way past maxStillnessAlerts)
        - [x] confirm that it logs to Sentry
    - [x] trigger maxStillnessAlerts + 1 stillness alerts
        - [x] confirm that it logs to Sentry
    - [x] trigger exactly maxStillnessAlerts stillness alerts
        - [x] confirm that it doesn't log to Sentry
    - [x] trigger like 1 stillness alert
        - [x] confirm that it doesn't log to Sentry
    - [x] don't trigger a stillness alert
        - [x] confirm it doesn't log to Sentry
- [x] do same series of tests as above but for a different interval (like 5 minutes to make my life easier)
- [ ]  SMOKE TESTS

### CODE REVIEW

**Comment by trisapeace** [comment]
Task linked: [CU-2chw8y0 Internal alert when a Sensors client or location is performing outside of a normal range](https://app.clickup.com/t/2chw8y0)

File: .travis.yml [hunk 1]
```diff
@@ -20,7 +20,7 @@ env:
     - SESSION_RESET_THRESHOLD_TEST=2000
     - SUBSEQUENT_ALERT_MESSAGE_THRESHOLD_TEST=1000
     - RADAR_WINDOW_SIZE_SECONDS_TEST=2
-    - LOW_BATTERY_ALERT_TIMEOUT_TEST=2
```
Comment (trisapeace) [inline .travis.yml:23]: Ah! I wonder if this is why the tests are failing in Travis, @susanchung-brave . I think you accidentally deleted this environment variable.

File: .travis.yml [hunk 2]
```diff
@@ -20,7 +20,7 @@ env:
     - SESSION_RESET_THRESHOLD_TEST=2000
```
Comment (trisapeace) [inline .travis.yml]: Friendly reminder to add a line in the CHANGELOG explaining what changed.

File: server/README.md [hunk 3]
```diff
@@ -110,7 +110,7 @@ then create it from the correct `.env` file (for example, on the Sensors Admin s
 
       - Version number = <version number>
 
-      - Title = v<version number> <envrionment>
+      - Title = v<version number> <environment>
```
Comment (trisapeace) [inline server/README.md:113]: Thanks for fixing all the typos!

File: server/.env.example [hunk 4]
```diff
@@ -101,6 +101,10 @@ ONESIGNAL_API_KEY_TEST=jumbleOfCharactersHere
 LOW_BATTERY_ALERT_TIMEOUT = 86400
 LOW_BATTERY_ALERT_TIMEOUT_TEST = 2
 
+# Max number of stillness alerts in 24 hours before it starts to become too frequent
```
Comment (trisapeace) [inline server/.env.example:104]: Thanks for adding the explanation!

File: server/db/db.js [hunk 5]
```diff
@@ -694,6 +694,33 @@ async function getLocationsFromClientId(clientId, pgClient) {
   }
 }
 
+async function numberOfStillnessAlertsIn24Hours(locationid, pgClient) {
+  const endDate = new Date()
+  const startDate = new Date(endDate.getTime() - 24 * 60 * 60 * 1000)
```
Comment (trisapeace) [inline server/db/db.js:699]: As we talked about on our call but that I'm mention here again so I don't forget to check for it again later... I would use the postgres command `NOW()` and the keyword `INTERVAL` to do this in the query itself because time is tricky and the server running the code may have a different time than the server running the DB.

File: server/db/db.js [hunk 6]
```diff
@@ -694,6 +694,33 @@ async function getLocationsFromClientId(clientId, pgClient) {
...
+  try {
+    const results = await helpers.runQuery(
+      'numberOfStillnessAlertsIn24Hours',
+      `
+      SELECT COUNT (*)
+      FROM sessions
+      WHERE alert_type='SENSOR_STILLNESS'
```
Comment (trisapeace) [inline server/db/db.js:706]: I think we have a constant somewhere that can be used instead of hardcoding the string `SENSOR_STILLNESS`. Generally speaking, it's better to use a constant than hardcoding values because it means that if something changes, then it only needs to change in one place.

File: server/index.js [hunk 7]
```diff
@@ -213,6 +213,7 @@ if (helpers.isTestEnvironment()) {
   }
   server = https.createServer(httpsOptions, app).listen(8080)
   setInterval(vitals.checkHeartbeat, 60000)
+  setInterval(vitals.checkForInternalProblems, 24 * 60 * 60 * 1000) // check every 24 hours to see if internal alerts need to be sent 24 * 60 * 60 * 1000, change to a shorter value for texting purposes
```
Comment (trisapeace) [inline server/index.js:216]: I wonder whether you want to make the 24 hours into an environment variable. It might make it easier for you to test because you'd be able to make it a smaller number in the dev environment. What do you think? If you do, then your DB function's name will need to change too so as not to reference 24 hours. Up to you!

File: server/sensor-helm-chart/templates/sensor-deployment.yaml [hunk 8]
```diff
@@ -165,6 +165,11 @@ spec:
               secretKeyRef:
                 name: {{ .Values.secretName | quote }}
                 key: CLICKUP_TEAM_ID
+          - name: MAX_STILLNESS_ALERTS
+            valueFrom:
+              secretKeyRef:
+                name: {{ .Values.secretName | quote }}
+                key: MAX_STILLNESS_ALERTS
```
Comment (trisapeace) [inline server/sensor-helm-chart/templates/sensor-deployment.yaml:172]: Yay! Remembered to do this!

When you deploy it on dev, we'll also need to remember to pull your branch on Sensors Admin. That was the other thing that we forgot last time.

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js [hunk 9]
```diff
@@ -0,0 +1,215 @@
+/* eslint-disable no-plusplus */
```
Comment (trisapeace) [inline server/test/unit/vitalsTest/checkForInternalProblemsTest.js:1]: Reminder that we spoke about moving this file into the `integration` folder :smile:

Comment (trisapeace) [inline server/test/unit/vitalsTest/checkForInternalProblemsTest.js:1]: Why did you disable the `no-plusplus` rule?

File: server/vitals.js [hunk 10]
```diff
@@ -121,6 +121,26 @@ async function checkHeartbeat() {
   }
 }
 
+// Check for internal issues by calling DB, if there is, then log error in Sentry
+async function checkForInternalProblems() {
+  const firmwareStateMachineLocations = await db.getActiveFirmwareStateMachineLocations()
```
Comment (trisapeace) [inline server/vitals.js:126]: I think I might have removed this function with the latest commit with the idea that `getLocations` could be used instead. Would that still work for you?

File: server/vitals.js [hunk 11]
```diff
@@ -121,6 +121,26 @@ async function checkHeartbeat() {
...
+  for (const location of firmwareStateMachineLocations) {
+    if (!location.client.isActive || !location.isActive) {
```
Comment (trisapeace) [inline server/vitals.js:128]: Note that this test will change when we do the rebase that will replace the `is_active` flag with `is_sending_alerts`, `is_sending_vitals`, and `is_displayed`. Which one(s) do you think we should use here?

File: server/vitals.js [hunk 12]
```diff
@@ -121,6 +121,26 @@ async function checkHeartbeat() {
...
+      continue
+    }
+    try {
+      const maxStillnessAlerts = parseInt(helpers.getEnvVar('MAX_STILLNESS_ALERTS'), 10)
```
Comment (trisapeace) [inline server/vitals.js:132]: I agree with you that this can be moved outside the `for` loop.

File: server/vitals.js [hunk 13]
```diff
@@ -121,6 +121,26 @@ async function checkHeartbeat() {
...
+      const numberOfStillnessAlerts = await db.numberOfStillnessAlertsIn24Hours(location.locationid)
+      const tooManyStillnessAlerts = numberOfStillnessAlerts > maxStillnessAlerts
+      if (tooManyStillnessAlerts) {
+        helpers.logSentry(`Unusually frequent number of stillness alerts have been received at ${location.locationid}`)
```
Comment (trisapeace) [inline server/vitals.js:136]: What do you think about adding the exact number of Stillness alerts in this message? Too much information?

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+
+// Configure Chai
+use(sinonChai)
+
+const sandbox = sinon.createSandbox()
+
+const maxStillnessAlerts = helpers.getEnvVar('MAX_STILLNESS_ALERTS')
+
+describe('vitals.js integration tests: checkForInternalProblems', () => {
+  /* eslint-disable no-underscore-dangle */
```
Comment (trisapeace) [inline server/test/unit/vitalsTest/checkForInternalProblemsTest.js:24]: Why did you need to disable `no-underscore-dangle`?

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    })
+
+    it('should not log any errors', () => {
+      expect(helpers.log).to.not.be.called
+    })
+  })
+
+  describe('should not log to Sentry if exactly maxStillnessAlerts have occurred in 24 hours at a location', () => {
+    beforeEach(async () => {
+      // Insert a single location and maxStillnessAlerts sessions + 1
```
Comment (trisapeace) [inline server/test/unit/vitalsTest/checkForInternalProblemsTest.js:72]: This commend doesn't seem to match what the test is doing. You actually have exactly `maxStillnessAlerts` sessions, right?

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+        clientId: client.id,
+        isActive: true,
+      })
+      for (let i = 1; i <= maxStillnessAlerts + 1; i++) {
+        this[`session${i}`] = await sessionDBFactory(db, {
+          locationid: this.testLocation.locationid,
+          alertType: ALERT_TYPE.SENSOR_DURATION,
+          incidentCategory: ALERT_TYPE.SENSOR_DURATION,
+        })
+        await helpers.sleep(1) // sleep to make sure that these have created_at values at least one millisecond apart
```
Comment (trisapeace) [inline server/test/unit/vitalsTest/checkForInternalProblemsTest.js:115]: Did you run into issues that this call to `helpers.sleep` fix?

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+    })
+
+    it('should not log any errors', () => {
+      expect(helpers.log).to.not.be.called
+    })
+  })
+
+  describe('duration alerts should not be counted, even if duration alerts exceed maxStillnessAlerts in 24 hours for a location', () => {
+    beforeEach(async () => {
+      // Insert a single location and maxStillnessAlerts sessions + 1
```
Comment (trisapeace) [inline server/test/unit/vitalsTest/checkForInternalProblemsTest.js:103]: Does this comment need updating?

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+      const client = await factories.clientDBFactory(db, { isActive: true })
+      this.testLocation = await locationDBFactory(db, {
+        clientId: client.id,
+        isActive: true,
+      })
+
+      this.incidentCategory = ALERT_TYPE.SENSOR_STILLNESS
+      this.alertType = ALERT_TYPE.SENSOR_STILLNESS
+      for (let i = 1; i <= maxStillnessAlerts + 1; i++) {
```
Comment (trisapeace) [inline server/test/unit/vitalsTest/checkForInternalProblemsTest.js:148]: Thanks for making this test at the boundary between it sending the alert and it not sending the alert.

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+          incidentCategory: this.incidentCategory,
+        })
+        await helpers.sleep(1) // sleep to make sure that these have created_at values at least one millisecond apart
+      }
+      await vitals.checkForInternalProblems()
+    })
+
+    it('should send one max stillness alert to Sentry', () => {
+      expect(helpers.logSentry).to.have.been.calledOnce
+    })
```
Comment (trisapeace) [inline server/test/unit/vitalsTest/checkForInternalProblemsTest.js:161]: Is this test necessary since you have the one right after it testing `it('should send one max stillness alert specific to the location that has exceeded the maxStillnessAlerts in 24 hours to Sentry'`?

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+      expect(helpers.logSentry).to.have.been.calledOnceWithExactly(
+        `Unusually frequent number of stillness alerts have been received at ${this.testLocation.locationid}`,
+      )
+    })
+
+    it('should not log any errors', () => {
+      expect(helpers.log).to.not.be.called
+    })
+  })
+})
```
Comment (trisapeace) [inline server/test/unit/vitalsTest/checkForInternalProblemsTest.js:215]: I'd also add test(s) around when there are Stillness Alerts outside the 24 hours.

**Review by trisapeace — COMMENTED** [review]
Looking good!

File: .travis.yml (hunk 2 above)
Comment (susanchung-brave) [reply .travis.yml]: Thank you, will do! : )

File: server/db/db.js (hunk 6 above)
Comment (susanchung-brave) [reply server/db/db.js:706]: That's very true! Thank you for pointing this out, I'll look for the constant

File: server/index.js (hunk 7 above)
Comment (susanchung-brave) [reply server/index.js:216]: Oooh yes I considered doing that, I probably will do that!

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js (hunk 9 above)
Comment (susanchung-brave) [reply server/test/unit/vitalsTest/checkForInternalProblemsTest.js:1]: I wanted to freely use ++ when I was writing my for loops for my tests 😄 I can definitely put it back though!

File: server/vitals.js (hunk 11 above)
Comment (susanchung-brave) [reply server/vitals.js:128]: I think we'd have to check `is_sending_vitals` to make sure it's sending vitals in the first place, and `is_sending_alerts` for both the location and the client?

File: server/vitals.js (hunk 10 above)
Comment (susanchung-brave) [reply server/vitals.js:126]: Thank you for pointing this out! I changed my code to use `getLocations` instead and I think it still works!

File: server/vitals.js (hunk 13 above)
Comment (susanchung-brave) [reply server/vitals.js:136]: I think my initial issue was that it'd be too much information, but thinking about it, I don't think it'd be a big deal if someone found out what we'd constitute as too many stillness alerts hahaha

It'll make our lives easier too if we see the exact number! Thank you for pointing this out!

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js (hunk 16 above)
Comment (susanchung-brave) [reply server/test/unit/vitalsTest/checkForInternalProblemsTest.js:115]: I assumed that I would when I put that in, but I didn't run into any issues taking it out! I think I left it in to be safe!

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js (hunk 17 above)
Comment (susanchung-brave) [reply server/test/unit/vitalsTest/checkForInternalProblemsTest.js:103]: Yes haha thank you for pointing that out!

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js (hunk 15 above)
Comment (susanchung-brave) [reply server/test/unit/vitalsTest/checkForInternalProblemsTest.js:72]: Oh yes! Thank you for pointing that comment out, I need to go fix it in my other tests haha

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js (hunk 19 above)
Comment (susanchung-brave) [reply server/test/unit/vitalsTest/checkForInternalProblemsTest.js:161]: That's a good point! I think I was thinking that if the second test fails, then I at least have this test to indicate that some type of message was received. But looking at the error messages I get, I think I can take it out!

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js (hunk 20 above)
Comment (susanchung-brave) [reply server/test/unit/vitalsTest/checkForInternalProblemsTest.js:215]: Really good point, thank you! I'll get to making those tests : )

File: server/test/unit/vitalsTest/checkForInternalProblemsTest.js (hunk 16 above)
Comment (susanchung-brave) [reply server/test/unit/vitalsTest/checkForInternalProblemsTest.js:115]: Actually I think it doesn't matter hahaha I'll go ahead and remove it!

File: server/test/integration/vitalsTest/checkForInternalProblemsTest.js [hunk 21]
```diff
@@ -0,0 +1,420 @@
...
+  describe('should not log to Sentry if exactly maxStillnessAlerts have occurred in set interval of time at a location', () => {
+    beforeEach(async () => {
+      // Insert a single location and maxStillnessAlerts sessions
+      const client = await factories.clientDBFactory(db, { isSendingAlerts: true })
+      this.testLocation = await locationDBFactory(db, {
+        clientId: client.id,
+        isSendingAlerts: true,
+      })
+
+      this.incidentCategory = ALERT_TYPE.SENSOR_STILLNESS
```
Comment (trisapeace) [inline server/test/integration/vitalsTest/checkForInternalProblemsTest.js:82]: `ALERT_TYPE.SENSOR_STILLNESS` isn't an incidentCategory. However, you don't really care what the incidentCategory is, so you can just leave it out and the sessionDBFactory will fill in a valid value.

File: server/test/integration/vitalsTest/checkForInternalProblemsTest.js [hunk 22]
```diff
@@ -0,0 +1,420 @@
...
+        `Unusually frequent number of stillness alerts (${maxStillnessAlerts + 1}) have been received at ${this.testLocation.locationid}`,
+      )
+    })
+
+    it('should not log any errors', () => {
+      expect(helpers.log).to.not.be.called
+    })
+  })
+
+  describe('alerts sent earlier than set interval of time prior should not be counted, even if the stillness alerts exceed maxStillnessAlerts', () => {
```
Comment (trisapeace) [inline server/test/integration/vitalsTest/checkForInternalProblemsTest.js:169]: The "even if the stillness alerts exceed maxStillnessAlerts" part of this test name makes me think that you this one _should_ send a Sentry alert and log an error. What is it that you wanted to test?

File: server/test/integration/vitalsTest/checkForInternalProblemsTest.js [hunk 23]
```diff
@@ -0,0 +1,420 @@
...
+          `Unusually frequent number of stillness alerts (${maxStillnessAlerts + 1}) have been received at location2`,
+        )
+      })
+
+      it('should not log any errors', () => {
+        expect(helpers.log).to.not.be.called
+      })
+    })
+  })
+})
```
Comment (trisapeace) [inline server/test/integration/vitalsTest/checkForInternalProblemsTest.js:420]: This whole thing is great. Very thorough testing. Thank you for doing all this :smile:

File: server/db/db.js [hunk 24]
```diff
@@ -13,7 +13,7 @@ const pool = new pg.Pool({
   user: helpers.getEnvVar('PG_USER'),
   database: helpers.getEnvVar('PG_DATABASE'),
   password: helpers.getEnvVar('PG_PASSWORD'),
-  ssl: { rejectUnauthorized: false },
+  ssl: false, // { rejectUnauthorized: false },
```
[the remark this answers is not in the document]
Comment (susanchung-brave) [reply server/db/db.js:16]: oops i can't believe i missed it on the very last push 😭 thanks for catching that!!

File: server/test/integration/vitalsTest/checkForInternalProblemsTest.js [hunk 25]
```diff
@@ -0,0 +1,420 @@
...
+const sinon = require('sinon')
+const sinonChai = require('sinon-chai')
+const rewire = require('rewire')
+
+// In-house dependencies
+const { ALERT_TYPE, factories, helpers } = require('brave-alert-lib')
+const db = require('../../../db/db')
+const { locationDBFactory, sessionDBFactory } = require('../../../testingHelpers')
+
+const vitals = rewire('../../../vitals')
```
[the remark this answers is not in the document]
Comment (susanchung-brave) [reply server/test/integration/vitalsTest/checkForInternalProblemsTest.js:13]: okay thanks!!

File: server/test/integration/vitalsTest/checkForInternalProblemsTest.js (hunk 21 above)
Comment (susanchung-brave) [reply server/test/integration/vitalsTest/checkForInternalProblemsTest.js:82]: oh that's very true, thank you for catching that! i think i got into the middle of changing it and then got distracted 😅

File: server/test/integration/vitalsTest/checkForInternalProblemsTest.js (hunk 22 above)
Comment (susanchung-brave) [reply server/test/integration/vitalsTest/checkForInternalProblemsTest.js:169]: i was trying to ensure that stillness alerts sent outside of the interval don't get counted, so that no sentry alert gets logged as a result! 

the scenario i wanted to avoid by making this test was having a stillness alert that was sent earlier than let's say 24 hours still contributing to the number of stillness alerts being counted in this period, which would falsely cause a sentry alert to be logged

i'll reword this to be more clear about it!

File: server/test/integration/vitalsTest/checkForInternalProblemsTest.js [hunk 26]
```diff
@@ -0,0 +1,420 @@
...
+      const client = await factories.clientDBFactory(db, { isSendingAlerts: true })
+      this.testLocation = await locationDBFactory(db, {
+        clientId: client.id,
+        isSendingAlerts: true,
+      })
+
+      this.incidentCategory = ALERT_TYPE.SENSOR_STILLNESS
+      this.alertType = ALERT_TYPE.SENSOR_STILLNESS
+      for (let i = 1; i <= maxStillnessAlerts; i += 1) {
+        this[`session${i}`] = await sessionDBFactory(db, {
```
[the remark this answers is not in the document]
Comment (susanchung-brave) [reply server/test/integration/vitalsTest/checkForInternalProblemsTest.js:85]: oh! thank you that is really clear! i'll do that right now!!

File: server/vitals.js [hunk 27]
```diff
@@ -123,6 +123,26 @@ async function checkHeartbeat() {
   }
 }
 
+// Check for internal issues by calling DB, if there is, then log error in Sentry
+async function checkForInternalProblems() {
+  const maxStillnessAlerts = parseInt(helpers.getEnvVar('MAX_STILLNESS_ALERTS'), 10)
+  const locations = await db.getLocations()
+  for (const location of locations) {
```
[the remark this answers is not in the document]
Comment (susanchung-brave) [reply server/vitals.js:130]: OH I should be checking that either the client isn't sending alerts OR the location itself isn't sending alerts right?

File: server/db/db.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (trisapeace) [reply server/db/db.js]: Haha, no worries. That's what fresh eyes are for!

File: server/test/integration/vitalsTest/checkForInternalProblemsTest.js [hunk 28]
```diff
@@ -0,0 +1,445 @@
...
+    it('should not send any alerts to Sentry', () => {
+      expect(helpers.logSentry).to.have.been.not.been.called
+    })
+
+    it('should not log any errors', () => {
+      expect(helpers.log).to.not.be.called
+    })
+  })
+
+  describe('should not log to Sentry if the location is not sending Alerts', () => {
```
Comment (trisapeace) [inline server/test/integration/vitalsTest/checkForInternalProblemsTest.js:99]: Nice! Good idea to add these two tests in.

**Review by trisapeace — COMMENTED** [review]
Looks fantastic.

Once you've deployed to dev and completed all the tests in your test plan, then you can "Squash & Merge"

**Review by trisapeace — APPROVED** [review]
Looks great. Once the smoke tests pass, then you can go ahead and Squash & Merge. Thanks for all your hard work on this one!
